radix-engine-utils/src/rocks_db_metrics/mod.rs

```diff
@@ -0,0 +1,544 @@
+use radix_engine_store_interface::interface::{
+    CommittableSubstateDatabase, DatabaseUpdates, DbPartitionKey, DbSortKey, DbSubstateValue,
+    PartitionEntry, SubstateDatabase,
+};
+use radix_engine_stores::{
+    memory_db::InMemorySubstateDatabase,
+    rocks_db::{BlockBasedOptions, Options, RocksdbSubstateStore},
+};
+use std::{cell::RefCell, collections::BTreeMap, path::PathBuf, time::Duration};
+
+/// Substate store with read time measurements for RocksDB and In Memory DB.
+pub struct SubstateStoreWithMetrics<S>
+where
+    S: SubstateDatabase + CommittableSubstateDatabase,
+{
+    db: S,
+    pub read_metrics: RefCell<BTreeMap<usize, Vec<Duration>>>,
+}
+
+impl SubstateStoreWithMetrics<RocksdbSubstateStore> {
+    pub fn new_rocksdb(path: PathBuf) -> Self {
+        let mut factory_opts = BlockBasedOptions::default();
+        factory_opts.disable_cache();
+
+        let mut opt = Options::default();
+        opt.set_disable_auto_compactions(true);
+        opt.create_if_missing(true);
+        opt.set_block_based_table_factory(&factory_opts);
+
+        Self {
+            db: RocksdbSubstateStore::with_options(&opt, path),
+            read_metrics: RefCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl SubstateStoreWithMetrics<InMemorySubstateDatabase> {
+    pub fn new_inmem() -> Self {
+        Self {
+            db: InMemorySubstateDatabase::standard(),
+            read_metrics: RefCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl<S: SubstateDatabase + CommittableSubstateDatabase> SubstateDatabase
+    for SubstateStoreWithMetrics<S>
+{
+    fn get_substate(
+        &self,
+        partition_key: &DbPartitionKey,
+        sort_key: &DbSortKey,
+    ) -> Option<DbSubstateValue> {
+        let start = std::time::Instant::now();
+        let ret = self.db.get_substate(partition_key, sort_key);
+        let duration = start.elapsed();
+
+        if let Some(value) = ret {
+            let exists = self.read_metrics.borrow().get(&value.len()).is_some();
+            if exists {
+                self.read_metrics
+                    .borrow_mut()
+                    .get_mut(&value.len())
+                    .unwrap()
+                    .push(duration);
+            } else {
+                self.read_metrics
+                    .borrow_mut()
+                    .insert(value.len(), vec![duration]);
+            }
+            Some(value)
+        } else {
+            None
```

Comment:
I think we should record both, no? My understanding is that RocksDB will need to check the level files before issuing a `None`.